### PR TITLE
MergeBlocks: Rewrite to use a generic algorithm

### PR DESCRIPTION
**Pull request by kripken**
Before this we had special logic for various call types. This replaces all
that with a single general code path, which unifies everything except
for control flow constructs (which remain as before, handled in a
special way for each of them).

The algorithm is simple and direct, basically it goes through the
children and when it finds a block, it sees if it can move the block's
contents outside of the parent. While doing so it takes into account
effects and so forth.

To make this easy, a random-access API is added to ChildIterator.

Diff without whitespace makes the existing test updates a lot simpler.
Note that this is not NFC as the old algorithm had some quirks like
not taking into account effects when there were more than 2 children;
the new code is uniform in how it handles things.

This ends up removing 19% of all blocks in j2wasm, which reduces
1% of total code size.

### CODE REVIEW

File: src/passes/ReorderFunctions.cpp [hunk 1]
```diff
@@ -24,7 +24,8 @@
 // a less beneficial position for compression, that is, mutually-compressible
 // functions are no longer together (when they were before, in the original
 // order, the has some natural tendency one way or the other). TODO: investigate
-// similarity ordering here.
+// similarity ordering here (see similarity-ordering branch and
+// https://www.rfk.id.au/blog/entry/cromulate-improve-compressibility/)
```
Comment (tlively) [inline src/passes/ReorderFunctions.cpp:28]: Intentional drive-by?

**Review by tlively — APPROVED** [review]
LGTM modulo those questions.

File: src/passes/MergeBlocks.cpp [hunk 2]
```diff
@@ -509,74 +509,152 @@ struct MergeBlocks
...
       }
-      outer = optimize(curr, curr->operands[i], outer);
-    }
-  }
 
-  void visitCall(Call* curr) { handleCall(curr); }
+      // Also give up if the block's last element has a different type than the
+      // block (leave that for refinalize).
+      auto* back = block->list.back();
+      if (block->type != back->type) {
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/MergeBlocks.cpp:581]: When replacing something with something else of a different type we'd need to run refinalize to update all the types etc. It's simpler to leave that to other passes to do beforehand. I'll add a comment.

File: src/passes/MergeBlocks.cpp [hunk 3]
```diff
@@ -509,74 +509,152 @@ struct MergeBlocks
...
+      // be helpful (we wouldn't actually be getting rid of the child block -
+      // although, in the binary format such blocks tend to vanish anyhow).
+      bool fail = false;
+      for (auto* blockChild : block->list) {
+        if (blockChild == back) {
+          break;
+        }
+        EffectAnalyzer blockChildEffects(
+          getPassOptions(), *getModule(), blockChild);
+        for (auto& effects : childEffects) {
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/MergeBlocks.cpp:602]: I considered that too. I'm not sure it's more efficient, as to add the effects we end up doing a merge operation. How much that saves later might make it worthwhile, but I'm not sure. (Perhaps we should add an API to scan and merge as a single operation?)

File: src/passes/ReorderFunctions.cpp (hunk 1 above)
Comment (kripken) [reply src/passes/ReorderFunctions.cpp:28]: Oh, sorry, yes, I guess I had this in my tree. I can remove it. I meanwhile opened #4322 anyhow.

**Pull request merged**